File: android/src/main/java/com/twiliovoicereactnative/ForkLogger.java
// FORK — KAR-878
// Owns: Twilio Voice native logging mirrored to the host application's Sentry SDK.
// Hooks into: SDKLog and native Android log bypasses.
// Re-check on SDK bump: Sentry Android breadcrumb/event APIs and SDKLog behavior.
package com.twiliovoicereactnative;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

import io.sentry.Breadcrumb;
import io.sentry.IScope;
import io.sentry.Sentry;
import io.sentry.SentryEvent;
import io.sentry.SentryLevel;

final class ForkLogger {
  private static final String CATEGORY = "twilio.voice.log";
  private static final String LOGGER_TAG = "twilio_voice.logger";
  private static final String REDACTED = "[REDACTED]";
  private static final Pattern JWT = Pattern.compile(
    "(?<![A-Za-z0-9_-])[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}(?![A-Za-z0-9_-])");
  private static final Pattern DEVICE_TOKEN = Pattern.compile(
    "(?i)((?:fcm|apns|device)(?: with)? token[=: ]{1,3})[a-z0-9:_-]{20,}");
  private static final Pattern NAMED_SECRET = Pattern.compile(
    "(?i)((?:access[_ -]?token|authorization|api[_ -]?key|secret|password)[=: ]{1,3})\\S+");

  private final String logTag;

  ForkLogger(@NonNull Class<?> source) {
    logTag = source.getSimpleName();
  }

  void debug(@NonNull String message) {
    if (!BuildConfig.DEBUG) return;

    final String safeMessage = redact(message);
    Log.d(logTag, safeMessage);
    addBreadcrumb(safeMessage, SentryLevel.DEBUG);
  }

  void info(@NonNull String message) {
    final String safeMessage = redact(message);
    Log.i(logTag, safeMessage);
    addBreadcrumb(safeMessage, SentryLevel.INFO);
  }

  void warning(@NonNull String message) {
    warning(null, message);
  }

  void warning(@Nullable Throwable cause, @NonNull String message) {
    final String safeMessage = redact(message);
    if (cause == null) {
      Log.w(logTag, safeMessage);
    } else {
      Log.w(logTag, safeMessage, cause);
    }
    capture(safeMessage, cause, SentryLevel.WARNING);
  }

  void error(@NonNull String message) {
    error(null, message);
  }

  void error(@Nullable Throwable cause, @NonNull String message) {
    final String safeMessage = redact(message);
    if (cause == null) {
      Log.e(logTag, safeMessage);
    } else {
      Log.e(logTag, safeMessage, cause);
    }
    capture(safeMessage, cause, SentryLevel.ERROR);
  }

  private void capture(@NonNull String message,
                       @Nullable Throwable cause,
                       @NonNull SentryLevel level) {
    addBreadcrumb(message, level);
    if (cause == null) {
      Sentry.captureMessage(message, level, scope -> enrichScope(scope, message, level));
      return;
    }

    final SentryEvent event = new SentryEvent(cause);
    event.setLevel(level);
    Sentry.captureEvent(event, scope -> enrichScope(scope, message, level));
  }

  private void addBreadcrumb(@NonNull String message, @NonNull SentryLevel level) {
    final Breadcrumb breadcrumb = new Breadcrumb(message);
    breadcrumb.setCategory(CATEGORY);
    breadcrumb.setLevel(level);
    breadcrumb.setData("logger", logTag);
    Sentry.addBreadcrumb(breadcrumb);
  }

  private void enrichScope(@NonNull IScope scope,
                           @NonNull String message,
                           @NonNull SentryLevel level) {
    scope.setTag(LOGGER_TAG, logTag);
    final Map<String, Object> context = new HashMap<>();
    context.put("logger", logTag);
    context.put("level", level.name().toLowerCase());
    context.put("message", message);
    scope.setContexts("twilio_voice_log", context);
  }

  @NonNull
  private static String redact(@NonNull String message) {
    final String withoutJwt = JWT.matcher(message).replaceAll(REDACTED);
    final String withoutDeviceToken = DEVICE_TOKEN.matcher(withoutJwt).replaceAll("$1" + REDACTED);
    return NAMED_SECRET.matcher(withoutDeviceToken).replaceAll("$1" + REDACTED);
  }
}

File: ios/ForkLogger.m
// FORK — KAR-878
// Owns: Twilio Voice native logging mirrored to the host application's Sentry SDK.
// Hooks into: Objective-C native log sites.
// Re-check on SDK bump: Sentry Cocoa breadcrumb/event APIs and native log sites.

@import Sentry;

#import "ForkLogger.h"

static NSString * const kForkLoggerCategory = @"twilio.voice.log";
static NSString * const kForkLoggerTag = @"twilio_voice.logger";
static NSString * const kForkLoggerName = @"TwilioVoiceReactNative";
static NSString * const kForkLoggerRedacted = @"[REDACTED]";

@implementation ForkLogger

+ (void)fork_debug:(NSString *)format, ... {
    va_list arguments;
    va_start(arguments, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:arguments];
    va_end(arguments);
    [self fork_emit:message level:kSentryLevelDebug cause:nil capture:NO];
}

+ (void)fork_info:(NSString *)format, ... {
    va_list arguments;
    va_start(arguments, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:arguments];
    va_end(arguments);
    [self fork_emit:message level:kSentryLevelInfo cause:nil capture:NO];
}

+ (void)fork_warning:(NSString *)format, ... {
    va_list arguments;
    va_start(arguments, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:arguments];
    va_end(arguments);
    [self fork_emit:message level:kSentryLevelWarning cause:nil capture:YES];
}

+ (void)fork_warningWithCause:(NSError *)cause format:(NSString *)format, ... {
    va_list arguments;
    va_start(arguments, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:arguments];
    va_end(arguments);
    [self fork_emit:message level:kSentryLevelWarning cause:cause capture:YES];
}

+ (void)fork_error:(NSString *)format, ... {
    va_list arguments;
    va_start(arguments, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:arguments];
    va_end(arguments);
    [self fork_emit:message level:kSentryLevelError cause:nil capture:YES];
}

+ (void)fork_errorWithCause:(NSError *)cause format:(NSString *)format, ... {
    va_list arguments;
    va_start(arguments, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:arguments];
    va_end(arguments);
    [self fork_emit:message level:kSentryLevelError cause:cause capture:YES];
}

+ (void)fork_emit:(NSString *)message
            level:(SentryLevel)level
            cause:(NSError *)cause
          capture:(BOOL)capture {
    NSString *safeMessage = [self fork_redact:message];
    NSLog(@"%@", safeMessage);

    SentryBreadcrumb *breadcrumb = [[SentryBreadcrumb alloc] initWithLevel:level
                                                                   category:kForkLoggerCategory];
    breadcrumb.message = safeMessage;
    breadcrumb.data = @{@"logger": kForkLoggerName};
    [SentrySDK addBreadcrumb:breadcrumb];

    if (!capture) return;

    void (^scopeBlock)(SentryScope *) = ^(SentryScope *scope) {
        [scope setTagValue:kForkLoggerName forKey:kForkLoggerTag];
        [scope setExtraValue:safeMessage forKey:@"twilio_voice_log_message"];
    };
    if (cause == nil) {
        [SentrySDK captureMessage:safeMessage withScopeBlock:^(SentryScope *scope) {
            [scope setLevel:level];
            scopeBlock(scope);
        }];
        return;
    }

    SentryEvent *event = [[SentryEvent alloc] initWithError:cause];
    event.level = level;
    [SentrySDK captureEvent:event withScopeBlock:scopeBlock];
}

+ (NSString *)fork_redact:(NSString *)message {
    NSString *safeMessage = message;
    NSArray<NSDictionary<NSString *, NSString *> *> *rules = @[
        @{
            @"pattern": @"(?<![A-Za-z0-9_-])[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}(?![A-Za-z0-9_-])",
            @"replacement": kForkLoggerRedacted,
        },
        @{
            @"pattern": @"(?i)((?:fcm|apns|device)(?: with)? token[=: ]{1,3})[a-z0-9:_-]{20,}",
            @"replacement": [@"$1" stringByAppendingString:kForkLoggerRedacted],
        },
        @{
            @"pattern": @"(?i)((?:access[_ -]?token|authorization|api[_ -]?key|secret|password)[=: ]{1,3})\\S+",
            @"replacement": [@"$1" stringByAppendingString:kForkLoggerRedacted],
        },
    ];

    for (NSDictionary<NSString *, NSString *> *rule in rules) {
        NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:rule[@"pattern"]
                                                                               options:0
                                                                                 error:nil];
        safeMessage = [regex stringByReplacingMatchesInString:safeMessage
                                                      options:0
                                                        range:NSMakeRange(0, safeMessage.length)
                                                 withTemplate:rule[@"replacement"]];
    }
    return safeMessage;
}

@end
